### DISCUSSION TITLE
Branding service boundaries

### DISCUSSION THREAD

**Discussion by tommasobertoni** [artiso-solutions/adsd-hotel#14] [General] [answered]
@artiso-solutions/systems-design 

During the analysis, we roughly described the `Branding` service as the owner of: room description, room picture, e-mail templates.
We can also include: UI styles, position of UI components...

But what about:
- _any_ other description (i.e. rate, room's additional services)
- names: should the services own the names of the "entities"? what about localization?

_If_ `Branding` owns how the data is shown to the users, does it also own the translations/localization?

___

<details>
  <summary>Let's assume it does.</summary>
<br/>

If `Branding` owns the names, than it also owns their storage.
The "entities" won't have a `name` property: the viewmodels will.

e.g. The `Blue` service, who owns the room types, doesn't need (care?) about the names shown to the users, and shouldn't be responsible for changing the value (or a translation of it) if the business decides to: `Branding` has that responsibility.

```csharp
class RoomType
{
    Guid Id { get; }
    int PeopleCapacity { get; }
    List<Bed> Beds { get; }
    ...
}
```

The UI gateway will then compose a viewmodel with the names and descriptions provided by `Branding`:
```csharp
var viewModels = new List<ViewModel>();
foreach (RoomType roomType in GetRoomTypes())
{
    var vm = AsViewModel(roomType);

    vm.Name = _brandingClient.GetRoomTypeName(roomType.Id);
    vm.Description = _brandingClient.GetRoomTypeDescription(roomType.Id);

    viewModels.Add(vm);
}
```

The `Branding` client (ambassador) can be culture-aware (device aware, context aware - _e,g. are we writing an email?_) and provide the appropriate value. This way, this awareness stops at the gateway and flows only through the `Branding` service.

**But then nobody can easily understand the data in the database!**

_Proposal_: allow entities to have an `InternalName` property, which *is not assumed to be shown to the user*, rather it's used as a human-readable "id". Something along the lines of [codenames](https://en.wikipedia.org/wiki/List_of_computer_technology_code_names).

</details>

___

<details>
  <summary>Let's assume it doesn't.</summary>
<br/>

Names and descriptions are part of the "entities". In order to have culture/device/context variants in all the services, `ITOps` could provide some helper functionalities for lookup operations.

Drawbacks:
- services must own the feature of providing variants
- services must store the variants (data ownership)
- requests to the services must include the culture/device/context info

</details>

**Comment by tziemek** [comment]
I'm in favor of 'Branding owns the data'. 
As for the understanding problem: Do we have to use a GUID as an identifier? We could use an *invariant* name as an identifier, i.e. your proposed InternalName could also be the identifier. And yes, we have to make sure they are unique.

**Comment by tommasobertoni** [comment, marked as answer]
As discussed in the [meeting](https://github.com/artiso-solutions/adsd-hotel/wiki/Meeting-notes#branding-service-boundaries-discussion), the `Branding` service will own the names, descriptions, additional info... that need to be viewed by the users (website, emails, bills...).

`Id`s will be the link between the services and `Branding` because they tend to be less change-prone compared to a "name" property.

It's not required that the services provide an `InternalName` property: only where it seems fit.

`Branding` will also own the format of this data, so both the following API definitions are valid:
```csharp
interface BrandingClient
{
    string GetRoomTypeName(string roomTypeId);
    string GetRoomTypeDescription(string roomTypeId);
}
```
```csharp
interface BrandingClient
{
    RoomTypeBranding GetRoomTypeBranding(string roomTypeId);
}

interface RoomTypeBranding
{
    string Name { get; }
    string Description { get; }
}
```

`Branding` will be responsible of providing the appropriate variant given the culture/device/context.
